Libraries/MLXLMCommon/ModelFactory.swift

```diff
@@ -0,0 +1,93 @@
+// Copyright © 2024 Apple Inc.
+
+import Foundation
+import Hub
+import Tokenizers
+
+public enum ModelFactoryError: Error {
+    case unsupportedModelType(String)
+    case unsupportedProcessorType(String)
+}
+
+/// Context of types that work together to provide a ``LanguageModel``.
+///
+/// A ``ModelContext`` is created by ``ModelFactory/load(hub:configuration:progressHandler:)``.
+/// This contains the following:
+///
+/// - ``ModelConfiguration`` -- identifier for the model
+/// - ``LanguageModel`` -- the model itself, see ``generate(input:parameters:context:didGenerate:)``
+/// - ``UserInputProcessor`` -- can convert ``UserInput`` into ``LMInput``
+/// - `Tokenizer` -- the tokenizer used by ``UserInputProcessor``
+///
+/// See also ``ModelFactory/loadContainer(hub:configuration:progressHandler:)`` and
+/// ``ModelContainer``.
+public struct ModelContext {
+    public let configuration: ModelConfiguration
+    public let model: any LanguageModel
+    public let processor: any UserInputProcessor
+    public let tokenizer: Tokenizer
+
+    public init(
+        configuration: ModelConfiguration, model: any LanguageModel,
+        processor: any UserInputProcessor, tokenizer: any Tokenizer
+    ) {
+        self.configuration = configuration
+        self.model = model
+        self.processor = processor
+        self.tokenizer = tokenizer
+    }
+}
+
+public protocol ModelFactory: Sendable {
```

Comment:
ModelFactory defines the API available for loading models.  LLM and VLM have concrete implementations because it needs to deal with the concrete types in those libraries (and they have slightly different internals, e.g. VLM needs to load additional configuration for the input processor)